### PR TITLE
FSE: Add domain-suggestions data store

### PR DESCRIPTION
**Pull request by sirreal**
Add `automattic/domains/suggestions` store script to FSE. This is a prerequisite for adding a domain picker in WP Admin.

See pbAok1-TR-p2

#### Changes proposed in this Pull Request

This PR adds a shared script to FSE. The script is called `a8c-fse-common-data-stores`. The script is built as part of the `common` bundle.

This necessitated a few changes to the webpack build:
- Allow multiple `source` arguments.
- Add special handling to webpack dependency extraction plugin.

In addition, a TypeScript definition has been added to the project to expose this "module". It will allow `import 'a8c-fse-common-data-stores';`.

The script is _registered_ in WordPress as part of the common bundle, but only enqueued as a script dependency when necessary.

This is important because these data stores should only be registered once (they should not be bundled in multiple scripts) and therefore should take advantage of this WordPress mechanism for sharing.

An alternative would be to rely on webpack dynamic imports, however that might require a major overhaul of the build system at this time because each package is built independently.

This is a first pass, scripts are only useful for their side effects. Further work would allow import/export from shared scripts if that were desired.

The only registered data store is `automattic/domains/suggestions` at this time.

#### Testing instructions

* Build and test with FSE - D43724-code can be used for easy testing.
* Sandbox `example.wordpress.com`
* Ensure that standard modules are externalized as before. You should see code like this in the build output:

```
/***/ "@wordpress/data":
/*!***************************************!*\
  !*** external {"this":["wp","data"]} ***!
  \***************************************/
/*! no static exports found */
/***/ (function(module, exports) {

(function() { module.exports = this["wp"]["data"]; }());

/***/ }),
```

* Add `import 'a8c-fse-common-data-stores';` to a module and ensure that `a8c-fse-common-data-stores` is added as a script dependency and causes the script to be enqueued.
* Test the data store, e.g.:
```js
wp.data.select( 'automattic/domains/suggestions' ).getCategories()
```

Note this will return `undefined` until it has resolved.

### CODE REVIEW

File: apps/full-site-editing/full-site-editing-plugin/common/data-stores/domain-suggestions.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+/**
+ * External dependencies
+ */
+import { DomainSuggestions } from '@automattic/data-stores';
+
+export const DOMAIN_SUGGESTIONS_KEY = DomainSuggestions.register( { vendor: 'variation2_front' } );
```
Comment (sirreal) [inline apps/full-site-editing/full-site-editing-plugin/common/data-stores/domain-suggestions.ts:6]: @daledupreez We'll be shipping this to WordPress.com Simple and Atomic sites. We definitely won't be able the Calypso file that sets the vendor variation there.

Do we go with something static (like the PR does right now)? We could use something on Simple sites that could be updated quickly and easily, but Atomic sites will be another matter 😞

Comment (daledupreez) [reply apps/full-site-editing/full-site-editing-plugin/common/data-stores/domain-suggestions.ts:6]: For now, let's stick with the current hard-coded value. Most of our testing will be during NUX, so shouldn't be needed for post-signup use cases.

File: apps/full-site-editing/webpack.config.js [hunk 2]
```diff
@@ -57,6 +71,24 @@ function getWebpackConfig( env = {}, argv = {} ) {
...
+		plugins: [
+			...webpackConfig.plugins.filter(
+				( plugin ) => plugin.constructor.name !== 'DependencyExtractionWebpackPlugin'
+			),
+			new DependencyExtractionWebpackPlugin( {
+				injectPolyfill: true,
+				requestToExternal( request ) {
+					if ( request.startsWith( FSE_MODULE_PREFIX ) ) {
+						switch ( request ) {
+							case 'a8c-fse-common-data-stores':
```
Comment (simison) [inline apps/full-site-editing/webpack.config.js:83]: Could have documentation pointing out where this name comes from since there's no convention of "handle matches a package in NPM"?

Comment (sirreal) [reply apps/full-site-editing/webpack.config.js:83]: This is the roughest spot here and I'm not sure what the best approach is. Documentation, at least, is a good start 👍

Comment (alshakero) [reply apps/full-site-editing/webpack.config.js:86]: Would it be possible to create an actual module that relies on the globally registered script?

A bit less magic-y.

Comment (alshakero) [reply apps/full-site-editing/webpack.config.js:86]: Would it be possible to create an actual module that relies on the globally registered script?

A bit less magic-y.

**Comment by sirreal** [comment]
> The PHP diff says "This large diff affects 332 files. Files without inline comments have been collapsed." Is this expected?

Thanks for checking. That's the complete build of FSE, it's expected.

File: apps/full-site-editing/full-site-editing-plugin/common/data-stores/domain-suggestions.ts (hunk 1 above)
Comment (simison) [reply apps/full-site-editing/full-site-editing-plugin/common/data-stores/domain-suggestions.ts]: @daledupreez Thanks! Do you see this is something we should deal with in future or are we good to do this and forget? 😁 

I'm thinking if this (and few other things I have in mind) would be good case to ask for more robust backend A/B testing tooling.

File: apps/full-site-editing/webpack.config.js (hunk 2 above)
Comment (sirreal) [reply apps/full-site-editing/webpack.config.js:86]: > Would it be possible to create an actual module that relies on the globally registered script?
> A bit less magic-y.

I'm generally in favor of reducing magic, but I'm not sure of a good way to do that here.

This quickly gets into some tricky webpack territory…

What would the import be? We have an actual module that we want to use, but we don't have a good way of referring to it. What are modules? Names registered on npm? Mapped aliases in webpack? Paths to files on local machines?

Ostensibly, we'd like to import the module that registers the data stores. It's in the repository and we can import it via relative imports, but not likely in a consistent way. With the current [`dependency-extraction-webpack-plugin` this may be impossible, I suspect if we were to change the `requestToExternal` API to provide the context](https://github.com/WordPress/gutenberg/blob/b2917b4906699a2085364fde4c6b9f6195618de4/packages/dependency-extraction-webpack-plugin/index.js#L45-L68), we might be able to determine that we're importing a specific file.

I believe this is all complicated by the fact that FSE JavaScript is output by a bunch of independent, unconnected webpack builds, so we need to stitch things together from outside. If we were to build everything together from a number of entrypoints, we might gain finer control over this sort of thing. But I'm speculating 🙂 

In short, this could be better and we can probable polish things or improve the conventions. I don't know a great way of doing that.

I _did_ add a faux module for TypeScript, so the typechecker will complain unknown modules are imported, and it understands that `a8c-fse-common-data-stores` is a module with no exports.

File: apps/full-site-editing/tsconfig.json [hunk 3]
```diff
@@ -2,6 +2,37 @@
 	"extends": "@automattic/calypso-build/tsconfig",
 	"exclude": [ "node_modules" ],
 	"compilerOptions": {
-		"allowJs": true
+		"target": "esnext",
```
Comment (noahtallen) [inline apps/full-site-editing/tsconfig.json:5]: are these rules not part of `"@automattic/calypso-build/tsconfig"`? seems like there are a lot of ways to have different TS configs in different packages or apps

Comment (sirreal) [reply apps/full-site-editing/tsconfig.json:5]: Yeah… improving TS setup across Calypso is a project in my backlog I haven't gotten to.

I'd like to deprecate, remove, and stop extending the tsconfig in calypso-build. I don't think there's much value there and the tsconfig should just be done per-project. There should be a base one specific to the Calypso monorepo that is extended for sub-projects.

Much of what's in here doesn't matter all that much. I believe it's only used by tsserver (editor integrations) and as long as it's providing good information, that's all we can ask 🙂 

There's almost no TypeScript in FSE, but I'm working on that 😁 Even then, we'll continue to use `calypso-build` for the foreseeable future which just strips out TypeScript constructs.

I'll remove the `extends` above.

File: apps/full-site-editing/tsconfig.json [hunk 4]
```diff
@@ -2,6 +2,37 @@
 	"extends": "@automattic/calypso-build/tsconfig",
```
Comment (sirreal) [inline apps/full-site-editing/tsconfig.json:2]: Don't extend calypso-build tsconfig.
```suggestion
```

**Comment by sirreal** [comment]
Rebased to fix conflict.

**Comment by yansern** [comment]
> * Ensure that standard modules are externalized as before. You should see code like this in the build output:

✅   Yes, I see them. 

image

> * Add `import 'a8c-fse-common-data-stores';` to a module and ensure that `a8c-fse-common-data-stores` is added as a script dependency and causes the script to be enqueued.

✅   For my testing, I added it to `full-site-editing-plugin/dotcom-fse/index.js`. I guess any JS files will work.

> * Test the data store, e.g.:
> ```js
> wp.data.select( 'automattic/domains/suggestions' ).getCategories()
> ```

✅  `getCategories()` doesn't exist in my testing (strange, wasn't this rebased?), but the domain stores is loaded.

image

**Comment by sirreal** [comment]
The FSE build failure is a known intermittent issue.

File: apps/full-site-editing/full-site-editing-plugin/common/data-stores/domain-suggestions.ts (hunk 1 above)
Comment (daledupreez) [reply apps/full-site-editing/full-site-editing-plugin/common/data-stores/domain-suggestions.ts]: We can leave this as-is for the time being. The real question here is that it's generally been easier to drive A/B tests from Calypso (without directly involving the back end), so that's how this has been done in the past.

If we're looking at a slower-changing client implementation in FSE, then it may make sense to change tack when we need to.
